File: Maratona_Pascal/OBI2014/setas.pas
(* Guilherme A. Pinto - OBI 2014 - Setas *)

program setas(input,output);
const
   MAXN	= 500;

var
   s	 : array [1..(MAXN+2),1..(MAXN+2)] of char;
   ans	 : longint;
   i,j,N : longint;

function solve(i, j :integer ): char;
var
   a,b : integer;
begin

  if ( (s[i,j] = 's') or (s[i,j] = '*') ) then
     solve := s[i,j]
  else begin
     a := i; b := j;
     if ( s[i,j] = 'A' ) then a := a-1;
     if ( s[i,j] = 'V' ) then a := a+1;
     if ( s[i,j] = '<' ) then b := b-1;
     if ( s[i,j] = '>' ) then b := b+1;
     s[i,j] := 's';
     s[i,j] := solve( a, b );
     if ( s[i,j] = '*' ) then ans := ans + 1;
     solve := s[i,j];
  end;
end;

begin
   readln(N);
   for j := 1 to N+2 do s[1,j] := '*';
   for i := 2 to N+1 do
      begin
	 s[i,1] := '*';
	 for j := 2 to N do read(s[i,j]);
	 readln(s[i,N+1]);
	 s[i,N+2] := '*';
      end;
   for j := 1 to N+2 do s[N+2,j] := '*';

   ans := 0;
   for i := 2 to N+1 do
      for j := 2 to N+1 do
	 if ( not (s[i,j] = '*') ) then
	    solve( i, j );

   writeln((N*N)-ans);
end.

File: Maratona_Java/src/OBI2014/setas.java
/* Guilherme A. Pinto - OBI 2014 - Setas */

import java.util.Scanner;
import java.io.*;

public class guilherme_java {

    static int MAX = 500;
    static char[][] s;
    static int ans;

    public static char solve( int i, int j ){
	int a,b;
	
	if ( s[i][j] == 's' || s[i][j] == '*' )
	    return s[i][j];
	
	a = i; b = j;
	switch( s[i][j] ){
	case 'A': a--; break;
	case 'V': a++; break;
	case '<': b--; break;
	case '>': b++; }
	
	s[i][j] = 's';
	s[i][j] = solve( a, b );
	if ( s[i][j] == '*' ) ans++;
	return s[i][j];
    }
    
    public static void main(String[] args) {
	Scanner in = new Scanner(System.in);
	int i,j,N;

	s = new char[MAX+2][MAX+2];

	N = in.nextInt(); in.nextLine();

	for ( i = 0; i < N+2; i++ ) s[0][i] = '*';

	for ( i = 1; i <= N; i++ ){
	    char[] buf = in.nextLine().toCharArray();
	    s[i][0] = '*';
	    for ( j = 0; j < N; j++ ) 
		s[i][j+1] = buf[j];
	    s[i][N+1] = '*';
	}
	for ( i = 0; i < N+2; i++ ) s[N+1][i] = '*';

	ans = 0;
	for ( i = 1; i <= N; i++ )
	    for ( j = 1; j<= N; j++ )
		if ( s[i][j] != '*' ) 
		    solve ( i, j );

	System.out.print(String.format("%d\n", (N*N)-ans ));
    }

}
